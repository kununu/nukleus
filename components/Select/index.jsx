// Read more about controlled components
// https://facebook.github.io/react/docs/forms.html#controlled-components
import React, {Component, PropTypes} from 'react';

import styles from './index.scss';

import {
  controlLabel,
  errorStyles,
  hidden,
  formControl,
  formGroup,
  srOnly
} from '../index.scss';


export default class Select extends Component {
  static propTypes = {
    autoFocus: PropTypes.bool,
    defaultItem: PropTypes.string,
    defaultRequired: PropTypes.string,
    disabled: PropTypes.bool,
    error: PropTypes.string,
    id: PropTypes.string.isRequired,
    inputStyle: PropTypes.string,
    items: PropTypes.object,
    labelHidden: PropTypes.bool,
    name: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    query: PropTypes.object,
    required: PropTypes.bool,
    title: PropTypes.string.isRequired,
    value: PropTypes.any
  };

  static defaultProps = {
    error: null,
    inputStyle: 'inline',
    query: {}
  };

  state = {
    showError: false,
    value: this.props.value || ''
  };

  componentWillMount () {
    this.updateValue(this.props.query[this.props.name] || this.props.value || '');
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.error) this.showError();
    if (!this.needsUpdate(nextProps)) return;
    this.updateValue(nextProps.query[this.props.name] || nextProps.value || '');
  }

  // Property initializer binds method to class instance
  onChange = (...args) => {
    this.updateValue(args[0].target.value);
    if (this.props.onChange) this.props.onChange(...args);
    this.hideError();
  };

  needsUpdate ({value, query}) {
    return (
      value !== this.props.value ||
      query !== this.props.query
    );
  }

  updateValue (value) {
    this.setState({value});
  }

  showError () {
    this.setState({showError: true});
  }

  hideError () {
    this.setState({showError: false});
  }

  render () {
    const {
      autoFocus,
      defaultItem,
      defaultRequired,
      disabled,
      error,
      id,
      inputStyle,
      items,
      labelHidden,
      name,
      required,
      title
    } = this.props;

    return (
      <div className={`${formGroup} ${styles[inputStyle]}`}>
        {labelHidden && <span className={srOnly}>{title}</span>}

        <label
          className={`${controlLabel} ${labelHidden && hidden}`}
          htmlFor={id}>{title}</label>

        <div className={styles.inputContainer}>

          <select
            name={name}
            value={this.state.value}
            id={id}
            required={required}
            autoFocus={autoFocus}
            onChange={this.onChange}
            className={`${formControl} ${styles.select}`}
            disabled={disabled}>

            {defaultRequired &&
              <option value="" hidden>{defaultRequired}</option>}

            {defaultItem &&
              <option value="">{defaultItem}</option>}

            {Object.keys(items).map(value =>
              <option
                key={value}
                value={value}>{items[value]}</option>
            )}
          </select>

          {this.state.showError && error &&
            <span className={errorStyles}>error</span>}
        </div>
      </div>
    );
  }
}
