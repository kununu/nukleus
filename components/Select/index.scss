@import '~kununu-theme/scss/base/variables';

.inline {
  @media (min-width: $screen-xs-min) {
    display: flex;
    flex-flow: row;
    align-items: center;
  }

  label {
    margin-right: 10px;
    margin-bottom: 0;
    width: 140px;
  }
}

.inputContainer {
  position: relative;
  width: 100%;

  * {
    box-sizing: border-box;
  }
}
