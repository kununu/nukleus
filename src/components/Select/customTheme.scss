@import '~@kununu/kununu-theme-v2/scss/base/variables';

.select {
  appearance: none;
  padding-right: 30px;
  padding-left: 30px;
  border-radius: 100px;
  border: 1px solid $green;
  box-shadow: inset 0 0 0 1px $green, inset 0 2px 1px rgba($black, .075);

  // Hide the arrow in IE10 and up
  &::-ms-expand {
    display: none;
  }

  &:focus {
    outline: none;
    border: 1px solid $green;
    box-shadow: inset 0 0 0 1px $green, inset 0 2px 1px rgba($black, .075);
  }

  // Disable the Firefox inner focus ring
  &:-moz-focusring {
    color: transparent;
    text-shadow: 0 0 0 $black;
  }
}

.innerSelectContainer {
  position: relative;
  width: 100%;

  * {
    box-sizing: border-box;
  }
}

.selectCaret {
  position: absolute;
  right: 20px;
  top: ($input-height-base / 2.7);
  z-index: 3;
  pointer-events: none;
  color: $green;

  svg {
    width: 12px;
    height: 12px;
  }

  path {
    fill: $green;
  }
}
