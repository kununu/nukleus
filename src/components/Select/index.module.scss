@import '~@kununu/kununu-theme-v2/scss/base/variables';

.select {
  appearance: none;
  padding-right: 30px;

  // Hide the arrow in IE10 and up
  &::-ms-expand {
    display: none;
  }

  // Disable the Firefox inner focus ring
  &:-moz-focusring {
    color: transparent;
    text-shadow: 0 0 0 $black;
  }
}

.innerSelectContainer {
  position: relative;
  width: 100%;

  * {
    box-sizing: border-box;
  }
}

.selectCaret {
  position: absolute;
  right: 15px;
  top: ($input-height-base / 2.5);
  font-size: 10px;
  z-index: 3;
  pointer-events: none;

  svg {
    width: 10px;
    height: 10px;
  }
}
