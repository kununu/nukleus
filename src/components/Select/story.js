import React from 'react';
import {storiesOf} from '@storybook/react';
import {defaultOptions} from 'defaultOptions'; // eslint-disable-line import/no-unresolved
import {withOptions} from '@storybook/addon-options';
import {action} from '@storybook/addon-actions';
import {withInfo} from '@storybook/addon-info';
import {
  withKnobs, text, boolean, select, object,
} from '@storybook/addon-knobs/react';

import ThemeProvider from '../ThemeProvider';

import Select from './index';
import customTheme from './customTheme.scss';

storiesOf('Select', module)
  .addDecorator(withOptions({...defaultOptions}))
  .addDecorator(withKnobs)
  .addDecorator((story, context) => withInfo('The select component!')(story)(context))
  .add('basics', () => (
    <div>
      <h3>Basic select with no selected value provided</h3>
      <div style={{position: 'relative'}}>
        <Select
          autoFocus={boolean('autoFocus', true)}
          label={text('label', 'Fruit')}
          labelHidden={boolean('labelHidden', false)}
          name={text('name', 'fruit1')}
          id={text('id', 'fruit1')}
          isRequired={boolean('isRequired', false)}
          inputStyle={select('inputStyle', {
            block: 'block',
            inline: 'inline',
          }, 'block')}
          onChange={action('onChange')}
          onBlur={action('onBlur')}
          onFocus={action('onFocus')}
          requiredLabel={text('requiredLabel', 'required')}
          items={object('items', {
            'option-1': 'Banana',
            'option-2': 'Apple',
            'option-3': 'Orange',
          })}
        />
      </div>

      <br />
      <br />
      <hr />
      <br />

      <h3>Select with a default item</h3>
      <p>The default item can be used when you want the user to be able to select nothing. The default item is an option tag without a value.</p>

      <div style={{position: 'relative'}}>
        <Select
          autoFocus={boolean('autoFocus', true)}
          defaultItem={text('defaultItem', 'Reselectable default item')}
          label={text('label', 'Fruit')}
          labelHidden={boolean('labelHidden', false)}
          name={text('name', 'fruit2')}
          id={text('id', 'fruit1')}
          isRequired={boolean('isRequired', false)}
          inputStyle={select('inputStyle', {
            block: 'block',
            inline: 'inline',
          }, 'block')}
          requiredLabel={text('requiredLabel', 'required')}
          onChange={action('onChange')}
          onBlur={action('onBlur')}
          onFocus={action('onFocus')}
          items={object('items', {
            'option-1': 'Banana',
            'option-2': 'Apple',
            'option-3': 'Orange',
          })}
        />
      </div>

      <br />
      <br />
      <hr />
      <br />

      <h3>Select with a default required item</h3>
      <p>
        The default required item will only be shown if nothing is selected and no value is given. Once a user
        selects an option the default required option disappears.
      </p>

      <div style={{position: 'relative'}}>
        <Select
          autoFocus={boolean('autoFocus', true)}
          defaultRequired={text('defaultRequired', 'Please select!')}
          label={text('label', 'Fruit')}
          labelHidden={boolean('labelHidden', false)}
          name={text('name', 'fruit2')}
          id={text('id', 'fruit1')}
          isRequired={boolean('isRequired', false)}
          inputStyle={select('inputStyle', {
            block: 'block',
            inline: 'inline',
          }, 'block')}
          requiredLabel={text('requiredLabel', 'required')}
          onChange={action('onChange')}
          onBlur={action('onBlur')}
          onFocus={action('onFocus')}
          items={object('items', {
            'option-1': 'Banana',
            'option-2': 'Apple',
            'option-3': 'Orange',
          })}
        />
      </div>
    </div>
  ))
  .add('With sorting', () => (
    <div>
      <p>Pass a sorting function to property sort to sort lists or objects without numbered keys.</p>
      <Select
        autoFocus={boolean('autoFocus', true)}
        label={text('label', 'Sorted fruit')}
        labelHidden={boolean('labelHidden', false)}
        name={text('name', 'fruit2')}
        id={text('id', 'fruit1')}
        isRequired={boolean('isRequired', false)}
        inputStyle={select('inputStyle', {
          block: 'block',
          inline: 'inline',
        }, 'block')}
        requiredLabel={text('requiredLabel', 'required')}
        onChange={action('onChange')}
        onBlur={action('onBlur')}
        onFocus={action('onFocus')}
        sort={(curr, next) => curr.value.localeCompare(next.value)}
        items={object('items', {
          'option-1': 'Plum',
          'option-2': 'Banana',
          'option-3': 'Apple',
          'option-4': 'Orange',
        })}
      />
    </div>
  ))
  .add('custom', () => (
    <ThemeProvider theme={customTheme}>
      <div style={{position: 'relative'}}>
        <Select
          autoFocus={boolean('autoFocus', true)}
          label={text('label', 'Fruit')}
          labelHidden={boolean('labelHidden', false)}
          name={text('name', 'fruit1')}
          id={text('id', 'fruit1')}
          isRequired={boolean('isRequired', false)}
          inputStyle={select('inputStyle', {
            block: 'block',
            inline: 'inline',
          }, 'block')}
          onChange={action('onChange')}
          onBlur={action('onBlur')}
          onFocus={action('onFocus')}
          requiredLabel={text('requiredLabel', 'required')}
          items={object('items', {
            'option-1': 'Banana',
            'option-2': 'Apple',
            'option-3': 'Orange',
          })}
        />
      </div>
    </ThemeProvider>
  ));
