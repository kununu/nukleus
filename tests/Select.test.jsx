import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';
import Select from 'Select'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies

const select = (
  <Select
    title="Select"
    name="select"
    id="select"
    isRequired
    value="option"
    items={{option: 'Option', 'option-2': 'Option 2', 'option-3': 'Option 3'}} />
);

test('Renders Select without crashing', () => {
  const component = renderer.create(select);
  expect(component.toJSON()).toMatchSnapshot();
});

test('Changes status on value change', () => {
  const component = mount(select);
  component.find({value: 'option-2'}).simulate('change');
  expect(toJson(component)).toMatchSnapshot();
});
