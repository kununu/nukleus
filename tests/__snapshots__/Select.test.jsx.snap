exports[`test Changes status on value change 1`] = `
<Select
  error={null}
  id="select"
  inputStyle="inline"
  items={
    Object {
      "option": "Option",
      "option-2": "Option 2",
      "option-3": "Option 3",
    }
  }
  name="select"
  query={Object {}}
  required={true}
  title="Select"
  value="option">
  <div
    className="formGroup inline">
    <label
      className="controlLabel undefined"
      htmlFor="select">
      Select
    </label>
    <div
      className="inputContainer">
      <select
        className="formControl select"
        id="select"
        name="select"
        onChange={[Function]}
        required={true}
        value="option-2">
        <option
          value="option">
          Option
        </option>
        <option
          value="option-2">
          Option 2
        </option>
        <option
          value="option-3">
          Option 3
        </option>
      </select>
    </div>
  </div>
</Select>
`;

exports[`test Renders Select without crashing 1`] = `
<div
  className="formGroup inline">
  <label
    className="controlLabel undefined"
    htmlFor="select">
    Select
  </label>
  <div
    className="inputContainer">
    <select
      autoFocus={undefined}
      className="formControl select"
      disabled={undefined}
      id="select"
      name="select"
      onChange={[Function]}
      required={true}
      value="option">
      <option
        value="option">
        Option
      </option>
      <option
        value="option-2">
        Option 2
      </option>
      <option
        value="option-3">
        Option 3
      </option>
    </select>
  </div>
</div>
`;
